filters/include/pcl/filters/functor_filter.h

```diff
@@ -76,22 +80,48 @@ class FunctorFilter : public FilterIndices<PointT> {
    */
   void
   applyFilter(Indices& indices) override
+  {
+    const auto exec = executor::best_fit{};
+    applyFilter(exec, indices);
+  }
+
+  template <typename Executor, typename executor::instance_of_base<Executor, executor::inline_executor, executor::omp_executor> = 0>
+  void
+  applyFilter (const Executor &exec, Indices& indices)
   {
     indices.clear();
-    indices.reserve(input_->size());
+    indices.reserve(indices_->size());
+    std::vector<std::uint8_t> keep(indices_->size(), 0);
+
     if (extract_removed_indices_) {
       removed_indices_->clear();
-      removed_indices_->reserve(input_->size());
+      removed_indices_->reserve(indices_->size());
     }
 
-    for (const auto index : *indices_) {
-      // functor returns true for points that should be selected
-      if (negative_ != functor_(*input_, index)) {
-        indices.push_back(index);
-      }
-      else if (extract_removed_indices_) {
-        removed_indices_->push_back(index);
+    auto cond_filtering = [&](executor::executor_index_t<Executor> index) {
+      pcl::utils::ignore(index);
+      #pragma omp for
+      for (index_t idx = 0; idx < indices_->size(); ++idx) {
+        exec.execute([&]() {
+          if (negative_ != functor_(*input_, idx)) {
+            keep[idx] = true;
+          }
+        });
       }
+    };
+
+    exec.bulk_execute(cond_filtering, 0);
+    merge_indices(exec, indices, keep);
+  }
+
+  template <typename Executor, typename executor::instance_of_base<Executor, executor::inline_executor, executor::omp_executor> = 0>
+  void merge_indices(const Executor &exec, Indices& indices, std::vector<std::uint8_t>& keep) {
```

Comment:
out parameters is a code-smell in modern C++. Construct and return `indices` by value inside this method. Trust RVO.
`keep` can be defined const.
definitely not a public method.

Reply:
That doesn't work for variables that might be holding a resource (like memory). out-parameters can reuse the allocated memory but new variable can't.

See: https://godbolt.org/z/s4Yf5d

Reply:
> variables that might be holding a resource (like memory)

We don't know how much space the user has reserved before hand.